cmd/zc_traverser_local.go

```diff
@@ -369,6 +427,194 @@ func (t *localTraverser) Traverse(preprocessor objectMorpher, processor objectPr
 		return fmt.Errorf("failed to scan path %s due to %s", t.fullPath, err.Error())
 	}
 
+	t.hashTargetChannel = make(chan string, 1_000) // "reasonable" backlog
+	// Use half of the available CPU cores for hashing to prevent throttling the STE too hard if hashing is still occurring when the first job part gets sent out
+	hashingThreadCount := runtime.NumCPU() / 2
+	hashError := make(chan error, hashingThreadCount)
+	wg := &sync.WaitGroup{}
+	commitMutex := &sync.Mutex{}
+	immediateStopHashing := int32(0)
+
+	finalizeHashing := func(existingErr error) error {
+		if existingErr != nil {
+			close(t.hashTargetChannel)                  // stop sending hashes
+			atomic.StoreInt32(&immediateStopHashing, 1) // force the end of hashing
+			wg.Wait()                                   // Await the finalization of all hashing
+
+			return existingErr // discard all hashing errors
+		} else {
+			close(t.hashTargetChannel) // stop sending hashes
+
+			wg.Wait()                    // Await the finalization of all hashing
+			close(hashError)             // close out the error channel
+			for err := range hashError { // inspect all hashing errors
+				if err != nil {
+					return err
+				}
+			}
+
+			return nil
+		}
+	}
+
+	// wrap the processor, preventing a data race
+	mutexProcessor := func(proc objectProcessor) objectProcessor {
+		return func(object StoredObject) error {
+			commitMutex.Lock()
+			defer commitMutex.Unlock()
+			err := proc(object)
+
+			return err
+		}
+	}
+
+	for i := 0; i < runtime.NumCPU()/2; i++ {
```

Comment:
we're duplicating here as well, you have already defined `hashingThreadCount` above, please use that instead